packages/flutter_tools/lib/src/custom_devices/custom_devices_config.dart

```diff
@@ -0,0 +1,90 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.8
+
+import 'package:meta/meta.dart';
+
+import '../base/config.dart';
+import '../base/file_system.dart';
+import '../base/logger.dart';
+import '../base/platform.dart';
+import 'custom_device_config.dart';
+
+/// Represents the custom devices config file on disk which in turn
+/// contains a list of individual custom device configs.
+class CustomDevicesConfig {
```

Comment:
Can you walk me through how you are loading/locating this config ?

Reply:
Of course. Generally, it's heavily inspired `PersistentToolState`.

- `CustomDevicesConfig` is just stored in the context. It's specified as part of the global context in `context_runner.dart`.
- when `CustomDevicesConfig` is instantiated, it creates an internal `Config` object which on construction loads the config file from a path determined by the `Config` object (using the config name, which is `custom_devices.json`) and parses it.
- that path is `%APPDATA%/.flutter_<config name>` on Windows and `$HOME/.config/flutter/<config name>` on MacOS and linux (see `Config._configPath`)

However, now I'm not so sure my initial description about the config file being automatically created when the custom devices feature is enabled is correct. Will the context runner call all the fallback generators eagerly? Or only when they're first used?

Reply:
They're only called when it is first used - I'm not so sure about having side effects like this in the generators.

How does a user discover how to add this config?

Reply:
> They're only called when it is first used - I'm not so sure about having side effects like this in the generators.

Yeah, I agree.

> How does a user discover how to add this config?

You mean, like, how should they know where the config is located? 

We could create the config when `flutter config --enable-custom-devices` is called and additionally let it output some message like `The custom devices configuration can now be edited at "C:\..."` so they have a path they can edit. (So it's both no side effects in generators and discoverable)

Reply:
Is that what you are doing for your device? WDUT about having a special command to add a device configuration?

Reply:
> Is that what you are doing for your device?

Not sure what you mean, I think there's nothing about this that is special to my device. That single path that'd be output by `flutter config --enable-custom-devices` is the only path any custom device configuration would ever be sourced from.

> WDUT about having a special command to add a device configuration?

Yeah we can do that. Actually, in the initial description I mentioned that a new custom devices config file, when generated, contains the schema tag and a disabled example device config. The point of the latter is so people can just copy+paste that entry in the config file to add their own device.

But yeah, I think a command would be cool as well, especially so its usable by scripts etc. (Or do IDE plugins use the daemon for that? not sure)

Reply:
The daemon probably wouldn't support this command but it would definitely be usable from the CLI. Doesn't need to be done in this PR, but imagine something like:

`flutter custom-device add /path/to/confg`
`flutter custom-device list`
`flutter custom-device remove name`

Reply:
Okay, but why would multiple config files be useful? Right now it's just a single, centralised config files that contain all custom devices to be configured (also mentioned that in the issue iirc)

Reply:
I would imagine those commands as add a configuration to the existing config file, instead of editing a file in the SDK right?

Reply:
Yeah I mean if you just want to add some custom device, editing the SDK config is better IMO. You've got a structure and an example config you can start with (and autocompletion bc of the JSON schema but thats of course not relevant when you're not using VS Code). If you start with a blank separate config file instead, even I would probably get some names / structures wrong and would start looking at the SDK config for guidance anyway. Especially since you'll probably configure the custom devices once and then completely forget about it, so not enough "exposure" to remember all the details

But if want to automate it & make it able to be used by scripts (or generally, non-interactively), then something like `flutter custom-device add <config>` is definitely useful, I'm just not sure that's the use-case you're caring about here.

I think a command like this would be more useful instead:
```
flutter custom-device add --ssh --hostname raspberrypi --user pi --launch "flutter-pi --observatory-host 192.168.178.43" --no-use-port-forwarding
```

That's just my POV though. We could of course also do both and your proposed config command would look like `flutter custom-device add --config /path/to/config` in that case.

Reply:
Understood. I think the current strategy is fine for now FWIW, and we can address any usability concerns in the future.

Reply:
I'll resolve this for now, I'll create an issue to track this